### PR TITLE
feat: ✨ add sd-header

### LINKED ISSUE #301: feat: ✨ add sd-header
## Description

The sd-header is meant to be overridable for the usage of its context.

## API

- prop `fixed`: `boolean` → adds shadow and adds fixed behaviour → currentHeight → :root → CSS Variable `--sd-header--height`

We should be able to do this:
```css
body{
padding: var(--sd-header--height);
}
```

- part: `main` to make everything overridable especially paddings and width → provide one responsive sample and non-responsive versions showing the single breakpoints

## Figma Comp 
https://github.com/solid-design-system/figma/issues/72

## Hand-off
https://www.figma.com/file/C0vgSK38CNYY83oAqh5s42/sd-header-V.0.0.1?type=design&node-id=10878-1611&mode=design&t=fNb9P9PYikurbXax-4

### Components detailed requirements
- [x] Props, CSS-Properties, Parts & Slots
  - [x] fixed boolean
- [x] Stories
  - [ ] tbd

## DoR
- [x] Item has business value
- [x] Item has been estimated by the team
- [x] Item is clear and well-defined
- [x] Item dependencies have been identified

## DoD
- [x] Documentation has been created/updated (if applicable)
- [x] Migration Guide has been created/updated (if applicable)
- [x] Relevant E2E tests (Features, A11y, Bug fixes) are created/updated
- [x] Relevant stories (Features, A11y) are created/updated
- [ ] Implementation works successfully on `feature` branch

### PR DESCRIPTION
**Pull request by azraefendic**
## Description:
Added sd-header component and samples. Closes #301 

## Definition of Reviewable:
- [x] Documentation has been created/updated (if applicable)
- [x] Migration Guide has been created/updated (if applicable)
- [x] Relevant E2E tests (Features, A11y, Bug fixes) are created/updated
- [x] Relevant stories (Features, A11y) are created/updated
- [x] Implementation works successfully on feature branch

### CODE REVIEW

File: packages/components/src/components/header/header.ts [hunk 1]
```diff
@@ -0,0 +1,158 @@
...
+    }
+  }
+
+  static styles = [
+    SolidElement.styles,
+    componentStyles,
+    css`
+      :host {
+        display: block;
+        z-index: 65536;
```
Comment (abudd1094) [inline packages/components/src/components/header/header.ts:122]: That's the most specific z-index I have ever seen 🤣

File: packages/components/src/components/header/header.ts [hunk 2]
```diff
@@ -0,0 +1,158 @@
...
+        position: fixed;
+      }
+
+      .fixed-shadow {
+        box-shadow: 0 4px 2px -2px gray;
+      }
+
+      [part='main'] {
+        margin: 0 auto;
+        width: var(--sd-header-inner-width, calc(100vw - 2 * var(--sd-header-padding-x, 16px)));
```
Comment (abudd1094) [inline packages/components/src/components/header/header.ts:138]: Aha, now I see, this is what makes the width overflow in the stories.  It looks strange in storybook but makes sense for the component.  Maybe we should just add a comment in the story to make it obvious for developers

**Review by abudd1094 — APPROVED** [review]
Meets requirements that I can see.  Tests passing, stories flowing.  Good work!

File: packages/components/src/components/header/header.ts (hunk 1 above)
Comment (azraefendic) [reply packages/components/src/components/header/header.ts:122]: Haha yes, I just copied the one already in use though 😁

File: packages/components/src/components/header/header.test.ts [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+    );
+    const shadowRoot = el.shadowRoot!;
+    const mainPart = shadowRoot.querySelector('[part="main"]')!;
+    const computedStyles = getComputedStyle(mainPart);
+
+    expect(computedStyles.width).to.equal('100px');
+    expect(computedStyles.maxWidth).to.equal('500px');
+    expect(computedStyles.paddingTop).to.equal('10px');
+    expect(computedStyles.paddingBottom).to.equal('20px');
+  });
```
Comment (mariohamann) [inline packages/components/src/components/header/header.test.ts:51]: Ui, wow!

File: packages/components/src/components/header/header.ts [hunk 4]
```diff
@@ -0,0 +1,158 @@
...
+ *
+ * @cssproperty --sd-header-inner-width - width of the header content
+ * @cssproperty --sd-header-inner-max-width - max-width of the header content
+ * @cssproperty --sd-header-padding-top - padding-top of the header content
+ * @cssproperty --sd-header-padding-bottom - padding-bottom of the header content
+ * @cssproperty --sd-header-padding-x - padding-left and padding-right of the header content
+ */
+@customElement('sd-header')
+export default class SdHeader extends SolidElement {
+  @property({ reflect: true, type: Boolean }) fixed = false;
```
Comment (mariohamann) [inline packages/components/src/components/header/header.ts:29]: Please document the props

File: packages/components/src/components/header/header.ts [hunk 5]
```diff
@@ -0,0 +1,158 @@
...
+  @property({ attribute: 'auto-spacing', reflect: true, type: Boolean }) autoSpacing = true;
```
Comment (mariohamann) [inline packages/components/src/components/header/header.ts:30]: booleans should always be false per default.

**Review by mariohamann — CHANGES_REQUESTED** [review]

File: packages/components/src/components/header/header.ts [hunk 6]
```diff
@@ -0,0 +1,158 @@
...
+      </header>
+    `;
+  }
+
+  private handleSlotChange(event: Event): void {
+    const slot = event.target as HTMLSlotElement;
+    if (slot.assignedElements().length > 0) {
+      this.refHeaderElement = event.currentTarget as HTMLElement;
+      this.setBodySpacing();
+      this.addResizeObserver();
```
Comment (mariohamann) [inline packages/components/src/components/header/header.ts:112]: I don't think that we need more then resizeObserver. Please use the same observer.

**Comment by mariohamann** [comment]
CleanShot 2023-11-03 at 15 41 57@2x

Please provide an example where the inner content has a max width

**Comment by coraliefeil** [comment]
Left comments on chromatic ... pls reassign when adjusted. Thank you <3

**Comment by coraliefeil** [comment]
Add fixed height for each viewport, pls.
Check shadow again, pls.
Reassign me when ready. THX

**Review by yoezlem — APPROVED** [review]

**Comment by karlbaumhauer** [comment]
As there have been so many reviews already, I will remove myself from the assignees. I dont feel, I could contribute anything meaningful to the PR and especially dont want to block the merge anymore.

**Review by mariohamann — APPROVED** [review]

**Pull request merged**